### PR TITLE
[BEAM-8972] Add Jenkins job with Combine test for portable Java

### PR DESCRIPTION
**Pull request by mwalenia**
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
retest this please

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
@lukecwik It seems that setting the Java home path from the CWiki broke the build - Gradle daemon experiences a context mismatch and doesn't work.

I'll change the paths back to working ones. 

Do you think we should add an Infra ticket to set env variables to java home paths? Or should we wait for Beam exclusive Jenkins and manage the envs there?

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by lukecwik** [comment]
Thinking about it more, the less that we have to configure manually within
Jenkins the better.
Just make them make them constants declared in one central location so it's
easy to change.

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
@tweise can you assist with diagnosing the errors in Flink? I don't know how to approach it

**Comment by mwalenia** [comment]
@mxm Would you be able to help with diagnosing problems plaguing the Flink cluster when running these tests?

**Comment by mxm** [comment]
Your root transform executes in an environment. This is not allowed in portability. We introduced READ support to implement cross-language support with existing sources, but it's only supported for streaming. For batch, you need to use the `BoundedReadViaImpulse` wrapper.

**Comment by lukecwik** [comment]
> Your root transform executes in an environment. This is not allowed in portability. We introduced READ support to implement cross-language support with existing sources, but it's only supported for streaming. For batch, you need to use the `BoundedReadViaImpulse` wrapper.

I removed `BoundedReadViaImpulse` in #10576. All Java read transforms are now expanded and executed as Splittable DoFns.

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Streaming

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mxm** [comment]
> > Your root transform executes in an environment. This is not allowed in portability. We introduced READ support to implement cross-language support with existing sources, but it's only supported for streaming. For batch, you need to use the `BoundedReadViaImpulse` wrapper.
> 
> I removed `BoundedReadViaImpulse` in #10576. All Java read transforms are now expanded and executed as Splittable DoFns as long as the `beam_fn_api` experiment is used.

@lukecwik I think we ought to auto-add the `beam_fn_api` flag in the Portable Runner similar as done in https://github.com/apache/beam/pull/9834/, this will avoid PRs like this one to manually add the flag. Tracking this here: https://jira.apache.org/jira/browse/BEAM-9538

File: .test-infra/jenkins/Flink.groovy [hunk 1]
```diff
@@ -17,8 +17,8 @@
  */
 
 class Flink {
-  private static final String flinkDownloadUrl = 'https://archive.apache.org/dist/flink/flink-1.10.0/flink-1.10.0-bin-scala_2.11.tgz'
-  private static final String hadoopDownloadUrl = 'https://repo.maven.apache.org/maven2/org/apache/flink/flink-shaded-hadoop-2-uber/2.8.3-9.0/flink-shaded-hadoop-2-uber-2.8.3-9.0.jar'
+  private static final String flinkDownloadUrl = 'https://archive.apache.org/dist/flink/flink-1.9.1/flink-1.9.1-bin-scala_2.11.tgz'
```
Comment (mxm) [inline .test-infra/jenkins/Flink.groovy]: Do you know what the problem was upgrading to Flink 1.10?

Comment (mwalenia) [reply .test-infra/jenkins/Flink.groovy]: The CLI of Yarn changed and no longer accepts one of options we supply. Kamil took care of it in #11146 - I'll remove this commit and rebase onto master in a while.

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Streaming

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Streaming

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Streaming

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Streaming

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Streaming

**Comment by mxm** [comment]
Unsubscribing but feel free to `@mxm` mention me.

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Streaming

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Streaming

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Streaming

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Streaming

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Streaming

**Comment by mwalenia** [comment]
@mxm I ran into a problem with TaskManagers timing out in a streaming case, the latest failed test is an example of this. Do you have any insight into why this might have happened?
Pinging @lukecwik as well.
Also there's an issue of intermittent `OutOfDirectMemoryError` getting thrown by gRPC on batch as well (eg. [here](https://builds.apache.org/job/beam_LoadTests_Java_Combine_Portable_Flink_Batch_PR/30/console) )

Thanks for your help!

**Comment by mxm** [comment]
>@mxm I ran into a problem with TaskManagers timing out in a streaming case, the latest failed test is an example of this. Do you have any insight into why this might have happened?

@mwalenia Not from the top of my head. I would need a stack trace or log to see what happened.

>Also there's an issue of intermittent OutOfDirectMemoryError getting thrown by gRPC on batch as well (eg. here)

> Also there's an issue of intermittent `OutOfDirectMemoryError` getting thrown by gRPC on batch as well (eg. [here](https://builds.apache.org/job/beam_LoadTests_Java_Combine_Portable_Flink_Batch_PR/30/console) )

Flink 1.10 configures has changed the memory configuration. It looks like GRPC's direct memory allocation is competing with Flink's managed memory. Flink allocates a fix amount of direct memory upfront leaving only a small amount left. See https://ci.apache.org/projects/flink/flink-docs-stable/ops/config.html#memory-configuration, especially `taskmanager.memory.framework.off-heap.size` and `taskmanager.memory.jvm-overhead.fraction`.

**Comment by mwalenia** [comment]
cc @kamilwu : Memory management changed in Flink 1.10, which might have some impact on Flink jobs we're seeing in tests.

**Comment by mwalenia** [comment]
@mxm I'm guessing this:
```
08:59:51 Exception in thread "main" java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.RuntimeException: The Runner experienced the following error during execution:
08:59:51 java.util.concurrent.TimeoutException: Heartbeat of TaskManager with id container_e01_1584690308819_0001_01_000002 timed out.
08:59:51 	at org.apache.beam.runners.portability.JobServicePipelineResult.waitUntilFinish(JobServicePipelineResult.java:98)
08:59:51 	at org.apache.beam.sdk.loadtests.LoadTest.run(LoadTest.java:99)
08:59:51 	at org.apache.beam.sdk.loadtests.CombineLoadTest.run(CombineLoadTest.java:66)
08:59:51 	at org.apache.beam.sdk.loadtests.CombineLoadTest.main(CombineLoadTest.java:169)
08:59:51 Caused by: java.util.concurrent.ExecutionException: java.lang.RuntimeException: The Runner experienced the following error during execution:
08:59:51 java.util.concurrent.TimeoutException: Heartbeat of TaskManager with id container_e01_1584690308819_0001_01_000002 timed out.
08:59:51 	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
08:59:51 	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1928)
08:59:51 	at org.apache.beam.runners.portability.JobServicePipelineResult.waitUntilFinish(JobServicePipelineResult.java:90)
08:59:51 	... 3 more
08:59:51 Caused by: java.lang.RuntimeException: The Runner experienced the following error during execution:
08:59:51 java.util.concurrent.TimeoutException: Heartbeat of TaskManager with id container_e01_1584690308819_0001_01_000002 timed out.
08:59:51 	at org.apache.beam.runners.portability.JobServicePipelineResult.propagateErrors(JobServicePipelineResult.java:165)
08:59:51 	at org.apache.beam.runners.portability.JobServicePipelineResult.waitUntilFinish(JobServicePipelineResult.java:110)
08:59:51 	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1604)
08:59:51 	at java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1596)
08:59:51 	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
08:59:51 	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
08:59:51 	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
08:59:51 	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
08:59:51 
08:59:51 > Task :sdks:java:testing:load-tests:run FAILED
```

is too little to diagnose anything. How can I find more logs from the failure? I can launch my own Flink cluster, but I'm not sure how to gather logs from TaskManagers.

**Comment by mxm** [comment]
If you run your own cluster, there is a log file for each task manager in the /logs directory.

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by mwalenia** [comment]
Run Load Tests Java Combine Portable Flink Streaming

**Pull request closed without merging**